lib/vmpooler/api/v1.rb

```diff
@@ -36,16 +36,46 @@ def need_token!
       validate_token(backend)
     end
 
-    def fetch_single_vm(template)
-      vm = backend.spop('vmpooler__ready__' + template)
-      return [vm, template] if vm
+    def pool_backend(pool)
+      pool_index = pool_index(pools)
+      pool_config = pools[pool_index[pool]]
+      backend = pool_config['clone_target'] || config['clone_target'] || 'default'
+      return backend
+    end
 
+    def fetch_single_vm(template)
+      template_backends = [template]
       aliases = Vmpooler::API.settings.config[:alias]
-      if aliases && aliased_template = aliases[template]
-        vm = backend.spop('vmpooler__ready__' + aliased_template)
-        return [vm, aliased_template] if vm
+      if aliases
+        template_backends << aliases[template] if aliases[template]
+
+        pool_index = pool_index(pools)
+        weighted_pools = {}
+        template_backends.each do |t|
+          next unless pool_index.key? t
+          index = pool_index[t]
+          clone_target = pools[index]['clone_target'] || config['clone_target']
+          next unless config.key?('backend_weight')
+          weight = config['backend_weight'][clone_target]
+          if weight
+            weighted_pools[t] = weight
+          end
+        end
+
+        if weighted_pools.count == template_backends.count
+          pickup = Pickup.new(weighted_pools)
+          selection = pickup.pick
+          template_backends.delete(selection)
```

Comment:
I think I'm following the logic now. Why not capture the selection here and in the `else` below. Then instead of doing `template_backends.each` on line 75 again, just `backend.spop()` the selection?

Reply:
My goal is to order the array so the first VM we try to retrieve is based on the selection from pickup. There is an each statement because we need to try this for the next one in the list in case there are no ready VMs for the first selection.